### PR TITLE
docs/rfcs: timeline ancestor detach API

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

When a tenant creates a new timeline that they will treat as their 'main' history,
it is awkward to permanently retain an 'old main' timeline as its ancestor. Currently
this is necessary because it is forbidden to delete a timeline which has descendents.

## Summary of changes

A new pageserver API is proposed to 'adopt' data from a parent timeline into
one of its children, such that the link between ancestor and child can be severed,
leaving the parent in a state where it may then be deleted.

### CODE REVIEW

File: docs/rfcs/033-ancestor-deletion.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+## Concurrent timeline creation
+
+If new historic timelines are created using the parent as an ancestor while the merge API is running, they will not be re-parented to the child. This doesn't break
+anything, but it leave the parent in a state where it might not be possible
+to delete it.
+
+Since timeline creations are an explicit user action, this is not something we need to
+worry about as the storage layer: a user who wants to delete their parent timeline will not create
+new children, and if they do, they can choose to delete those children to
+enable deleting the parent.
```
Comment (koivunej) [inline docs/rfcs/033-ancestor-deletion.md:231]: This does not concern the proposed merge_parent functionality however, or at least I cannot see it. For example an automation could continue branching from the head of the "old main" while this operation is ongoing, and that'd be no issue.

File: docs/rfcs/033-ancestor-deletion.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
```
Comment (koivunej) [inline docs/rfcs/033-ancestor-deletion.md:232]: There could be a concurrent gc which removes the "old main" layers up to the one branch point while we are adopting. This problem can be solved by having the remote timeline client of "old main" do the copying in order before unlinking the layers.

The "merge_parent" operation keeping a hold of the adopted layers will inhibit their actual removal from remote storage but not their unlinking. That will create a complication in case we need to restart while the operation is ongoing and those layers would become inaccessible *after* restart.

Well, it sounds quite pathological case of "fast layer upload" and "slow remote copy".

File: docs/rfcs/033-ancestor-deletion.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+## `merge_parent` API implementation
+
+Terms used in the following sections:
+
+- "the child": the timeline whose ID is specified in the merge API URL, also
+  called "new main" in the example.
+- "the parent": the parent of "the child". Also called "old main" in the example.
+- "the branch point" the ancestor_lsn of "the child"
+
```
Comment (koivunej) [inline docs/rfcs/033-ancestor-deletion.md:128]: Is there a "Phase 0: create a branch to inhibit gc as with any branching operation"

I think you are alluding to this by "the branch point" referring to **ancestor_lsn** but just to make sure I understand this correctly.

File: docs/rfcs/033-ancestor-deletion.md [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+into the child's IndexPart and upload to S3. This may be done while the
+child Timeline is still running.
+
+### Phase 3: modify timelines ancestry
+
+Modify the child's ancestor to None, and upload its IndexPart to persist the change.
+
+For all timelines which have the same parent as the child, and have a branch
+point lower than our branch point, switch their ancestor_timeline to the child,
+and upload their IndexPart to persist the change.
```
Comment (koivunej) [inline docs/rfcs/033-ancestor-deletion.md:161]: For this phase we'll need some gc inhibition as well; would be the easiest if we wouldn't activate the timeline before all reparting has been done so there is no worry about gc.

**Review by koivunej — APPROVED** [review]
This makes sense, and I agree with perf analysis and alternatives.

File: docs/rfcs/033-ancestor-deletion.md (hunk 2 above)
Comment (koivunej) [inline docs/rfcs/033-ancestor-deletion.md:232]: For consumption metrics, if we'd just copy the logical size cache while when unlinking the "old main" and "new main", we would get disruptionless synthetic size.

The synthetic size will then increase a lot temporarily because of multiple root timelines. I think that is fine. Or has there been any discussion that it shouldn't increase? If there should not be any billed cost we could have a "weak_ancestor" on the "new main" pointing to "old main", and reducing the cost of the duplication operation.

File: docs/rfcs/033-ancestor-deletion.md (hunk 4 above)
Comment (jcsp) [reply docs/rfcs/033-ancestor-deletion.md:161]: You're right -- I think the critical point to inhibit GC is between the update to the child's ("new main") index, and the point where we assign the other timelines to it as children.

File: docs/rfcs/033-ancestor-deletion.md (hunk 1 above)
Comment (jcsp) [reply docs/rfcs/033-ancestor-deletion.md:231]: Right, that's kind of what I'm saying here: if new timelines are created then we aren't doing anything bad, it's just... probably not what the user meant to do.

File: docs/rfcs/033-ancestor-deletion.md (hunk 2 above)
Comment (jcsp) [reply docs/rfcs/033-ancestor-deletion.md:232]: Interesting to think about.  They will indeed have a moment when the synthetic size ~doubles, before some higher layer deletes the old main timeline.

My bet is that since we charge per `size/time`, so a short spike probably won't show up in a way that upsets anyone, and we're kind of accurately passing along our costs to store both timelines for a moment.  Still, it would be kind of unfair if the charge appears for some users and not for others based on whether a consumption metrics poll hits the system at the right moment.

I think our starting point can be to do no special handling, and if we wanted to later change that we could add a flag to timelines to disable counting them for consumption metrics, and add an option to the merge API that would automatically set the "free" flag on the parent.  Something to pick up when integrating this API with cloud control plane.

File: docs/rfcs/033-ancestor-deletion.md [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+As an optimization, if there are no other timelines that will be adopted into
+the child, _and_ the child's image layers already full cover the branch LSN,
+then we may skip adopting layers.
+
+### Phase 2: update the child's index
+
+Having written out all needed layers in phase 1, atomically link them all
+into the child's IndexPart and upload to S3. This may be done while the
+child Timeline is still running.
```
Comment (koivunej) [inline docs/rfcs/033-ancestor-deletion.md:153]: This does leave us in a pickle:

We would continue to deny requests to delete the "old main" timeline while these would exist.

At runtime we don't have a way to change ancestor right now, but I cannot see why not an sync RwLock could be used there (`Timeline::ancestor_timeline`)... An RwLock would complicate all of the operations, and for example cancel/shutdown propagation from ancestor `Timeline::get` to reparented child `Timeline::get`, but that situation is pathological as well.

Interestingly the `Timeline::get_ancestor_timeline_id` is more used than `Timeline::ancestor_timeline` or `Timeline::get_ancestor_timeline`. Well, after a quick clance I dont think there are any major issues.

**Comment by stepashka** [comment]
added @ololobus and @prepor here to reviewers as representatives of expected consumers of the new API

File: docs/rfcs/033-ancestor-deletion.md [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+of the order 10-100ms, which is the time taken to upload their IndexPart, and
+restart the Timeline.
+
+# Interaction with other features
+
+## Concurrent timeline creation
+
+If new historic timelines are created using the parent as an ancestor while the merge API is running, they will not be re-parented to the child. This doesn't break
+anything, but it leave the parent in a state where it might not be possible
+to delete it.
```
Comment (arpad-m) [inline docs/rfcs/033-ancestor-deletion.md:226]: Inability to delete is a breakage from the end user point of view, as it will likely result in the tenant becoming stuck because the deletion of the old parent failed. Below you suggest the following:

> they can choose to delete those children to enable deleting the parent.

We likely do not want to delete that timeline because it might potentially have had writes.

I'd also like some discussion about race conditions, say if timeline creation is spawned in the background (currently that doesn't happen but it might in the future) and it isn't finished yet by the time the merge API endpoint is called, so it doesn't appear as child for the merge.

For example the RFC could say that callers of the pageserver APIs need to ensure that all creations have finished before a merge is conducted.

**Comment by arpad-m** [comment]
Maybe the RFC could have a section about interaction with WAL DR in the interaction with other features section?

WAL DR currently only supports the main branch but with the data that the safekeepers have, it can be extended to support DR also for tenants with branches.

If we delete entire timelines however, then we also delete the information whose timeline's WAL to use or copy. And if the old root timeline is being deleted, then one also deletes the initdb archive. Both would be required for building a WAL DR feature supporting merged branches in the future.

One way would be to keep a record file of historic timeline IDs so that we can ask safekeepers for WAL from those timelines. For the initdb we could copy it over, but of course the initdb won't cleanly apply with the WAL of the merged root timeline, one would have to apply it together with the original timeline's WAL.

Still, the merge needs to happen after we have implemented retaining both types of information in order for us to be able to WAL DR, regardless of when that WAL DR for branching feature is implemented.

In other words, we need to decide whether to implement retaining of this information, or whether it's okay to have a bunch of timelines that we really won't be able to WAL DR.

**Comment by prepor** [comment]
I think, from Neon Cloud perspective the definition/motivation is a little bit different. Let me try to formulate what we need as I see it, probably it'll change the final solution.

Let's consider the restore use case:

1. initial state
```
--- main branch (timeline 1) --->
```

2. we are restoring branch to the point `a`

```
-a- old branch (timeline 1) --->
 \
  - main branch (timeline 2) ->
```

3. after checking that all is fine user would want to delete the old branch because it isn't needed anymore. But he/she can't because it's now the branch with children.

4. With each and every new restoration the hierarchy will grow, and the number of layers will increase.

Another case with branch reset to parent:

1. initial state
```
-a- main branch (timeline 1) --->
 \
  - child branch (timeline 2) -b->
                                \
                                 - grandchild branch (timeline 3) -->
```

2. resetting of child branch leads to:
```
-a- main branch (timeline 1) --->
 \                               \
  \                               - child branch (timeline 4) -->
   \- old branch (timeline 2) -b->
                               \
                                - grandchild branch (timeline 3) -->

```

And again, we can't delete "old branch" because "grandchild" is pointing to it.


From UX perspective we can just hide such branches from the user after "deleting" them. But instead of actually deleting we can mark them as "abandoned" and exclude from UI.

But such branches will continue to exist on pageserver level and eat resources. That's how I see the definition of the problem. Instead of "merging" timelines, we need to implement a new type of Garbage Collecting. 

API (draft) that I'm proposing is consisting of two parts:

- GET /abandoned
The list of abandoned timelines. Pageserver pulls this list and merges timelines, deletes non needed parts or even whole timeline.
- POST /optimized
After finishing, pageserver notifies Control Plane about results: which timelines were merged to which and which were completely deleted. Control Plane modifies it's hierarchy accordingly.

We also can implement "normal" timeline/tenant deletion via this way.

What I like about this API:
- pull API: pageserver decides when it has resources for merging and controls rate
- it can execute operations in batches, grouped by tenant, for example
- we do not block any other operations for user: deleting / merging is a background process.

**Comment by jcsp** [comment]
>  API (draft) that I'm proposing is consisting of two parts:
>GET /abandoned

The "abandoned" timeline can have many children, so the pageserver needs to know which one to merge into.  Look at the "Before" example in the RFC to understand why -- in that case, simply indicating that the "old main" is abandoned would result in ambiguity.

The point about push vs. pull APIs is something that perhaps can be handled at the control plane layer if it is important for the console to avoid having to call out to the pageserver.  For the pageserver to continuously poll the control plane for work doesn't make sense to me: the control plane can internally poll its database much more efficiently.

**Comment by prepor** [comment]
> The "abandoned" timeline can have many children, so the pageserver needs to know which one to merge into.

That's the whole point of my proposal. Instead of "command" pageserver what to do, control plane could just notify pageserver that it doesn't need this timeline anymore, so pageserver can make some optimizations if it wants/is able. Pageserver has all the information about the timeline's children and their branching points.

> the control plane can internally poll its database much more efficiently.

polling database is cheap, merging / deleting resources is expensive. so we can leave the decision on when to do it to pageserver. Even if the logic of this decision initially could be very simple, we can later optimize it and even put it, for example, into a separate asynchronous process. And all of it without changing API and without blocking any other operations for the user.

**Comment by jcsp** [comment]
Notes from chat:
- The Console concept is to "abandon" a timeline: this means that a user-facing Branch no longer refers to the timeline.
- In the RFC example (child a, child b etc), marking "old main" as abandoned would only cause the  Timeline to be truncated to the LSN where child B branches off.
- In the same example, if child B was marked abandoned, then one could proceed to do a merge of new main into old main, and re-parent child A to new main.
- So there are two primitive operations: merging timeline into a parent timeline (as already described in the RFC), and truncating a timeline.  Some extra logic (either in PS or CP) would analyze the graph of Timelines (and whether they are abandoned), and figure out which Timelines need to be merged or truncated.
- A still-existing Branch's Timeline might have an abandoned Timeline as its ancestor, in which case requests to create new Branches from before its Timeline's branch point would need to be directed to the abandoned parent.  The take from the storage side is that this ancestor-recursing logic belongs in control plane, and that we can diverge the semantics of branches and timelines, such that:
  - For a Timeline, one may only create a child from an LSN within that timeline (i.e. not before its branch point)
  - For a Branch, one may create a child from any LSN in its history, and the actual Timeline ancestor of that created branch may be different than the Timeline of the Branch ancestor.
- The question of push vs. pull APIs is something perhaps for the CP team, as they would either way own the mapping of abandon UPDATEs into Operations, or providing an API for the pageserver to poll.  The inclination in storage is to stick with an imperative API to match how timeline creation/deletion is done.

**Comment by ololobus** [comment]
I went through the RFC and discussion and like Andrey's examples, they seem to be more real-life at this point. Not fully share the alternative API proposal, though, as I don't really like this bi-directional reconciliation kind of stuff

> The Console concept is to "abandon" a timeline: this means that a user-facing Branch no longer refers to the timeline

It's not an existing concept and I'm not 100% we need it for real

Right now, in reset API we have an option for keeping or deleting the previous branch. Due to the limitations this RFC is intended to resolve, we just forbid reset without keeping the old branch if it has ancestors. Like the Andrey's example

```
-a- main branch (timeline 1) --->
 \
  - child branch (timeline 2) -b->
                                  \
                                   - grandchild branch (timeline 3) -->
```

so we do this instead
```
-a- main branch (timeline 1) --->
 \                               \
  \                               - child branch (timeline 4) -->
   \
    - old branch (timeline 2) -b->
                                  \
                                   - grandchild branch (timeline 3) -->
```

What do we expect after implementing this RFC?

My understanding is that we again have two options:
1. User decided to keep previous branch -> OK, we do it as before
2. User decided to delete it right away -> then, I think we just use new API to merge timeline 2 into timeline 3 and end up in a state
```
-a- main branch (timeline 1) --->
 \                               \
  \                               - child branch (timeline 4) -->
   \
    - grandchild branch (timeline 3) -->
```
when timeline 3 is a direct child of timeline 1, right?

I think case 1. with user deciding to delete old branch after that falls into the same category as 2.

In 2., what if old branch has a full-blown hierarchy of children? IIUC, cplane behavior will be to just figure out all direct children (top common ancestors) of old branch aka timeline 2 and merge them with timeline 1. Will pageserver be happy with such a bulk-merging of 10, 20, 100 branches? If not, what are we going to do instead?

At first glance I don't really like this

> A still-existing Branch's Timeline might have an abandoned Timeline as its ancestor, in which case requests to create new Branches from before its Timeline's branch point would need to be directed to the abandoned parent. 

because it creates two trees: user-visible and a real shadow one, which will make it harder to explain to user what occupies the space. And with solution above we won't need to keep abandoned timelines at all

**Comment by ololobus** [comment]
> A still-existing Branch's Timeline might have an abandoned Timeline as its ancestor, in which case requests to create new Branches from before its Timeline's branch point would need to be directed to the abandoned parent.

If we want to keep such hidden timelines, then we don't really need the proposed merge API, I suppose. Instead, we just need to trim old branch aka timeline 2 to the LSN == max(branching LSN of all existing children, timeline 3 in this case), so that it doesn't occupy any extra space, only needed for copy-on-write to work

**Comment by jcsp** [comment]
>If we want to keep such hidden timelines, then we don't really need the proposed merge API, I suppose. 

We still need it: if someone uses the reset API repeatedly we must avoid a situation where they end up with a deeply nested tree of branches.

Keeping old timeline is only useful if we want to keep child branches that have a branch point after the LSN we reset to, and avoid the storage cost of having two totally separate histories (the "totally separate histories" was part of the original RFC proposal, on the basis that this would be a transient state and user would soon delete their old branches).

I think we have a choice between:
A) a more general solution (abandoned branches), where a user can arbitrarily select which branches are unwanted, and have our system do the merging/truncating as needed.
B) and a more special case solution (promoting a specific branch to "new main", which makes it standalone and makes it take ownership of any children before its branch point).

The branch-merging (aka "ancestor branch deletion", since merging is an overloaded term) is common work between them: that's the thing I think it makes sense to proceed with building, while the discussion about how higher layers will expose these concepts to users continue.

File: docs/rfcs/033-ancestor-deletion.md (hunk 6 above)
Comment (koivunej) [reply docs/rfcs/033-ancestor-deletion.md:226]: This is a rather interesting dimension, as in, should there be a reparenting API as well. Theoretically we could do reparenting by hand, but in practice not that easily.

> For example the RFC could say that callers of the pageserver APIs need to ensure that all creations have finished before a merge is conducted.

This sounds good for me.

Comment (koivunej) [reply docs/rfcs/033-ancestor-deletion.md:226]: [39054a6](https://github.com/neondatabase/neon/pull/6888/commits/39054a6191eb3a5170f2609a0b3dbb0d2fb62058)

**Review by arpad-m — APPROVED** [review]
the points I raised are sufficiently addressed

File: docs/rfcs/033-ancestor-deletion.md (hunk 5 above)
Comment (koivunej) [reply docs/rfcs/033-ancestor-deletion.md:153]: Attempted the above idea in #7285.

File: docs/rfcs/033-ancestor-deletion.md [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+wish to detach from its parent.
+
+On success, this API will leave the following state:
+
+- The detached child timeline will no longer have an ancestor, and will contain all
+  the data needed to service reads without recursing into an ancestor.
+- Any other children of the parent whose timeline points were at a lower LSN than
+  the detached child timeline will be modified to have the child timeline as their
+  new parent.
+- The parent timeline will still exist, but the child will no longer have it as an
```
Comment (stepashka) [inline docs/rfcs/033-ancestor-deletion.md:39]: what will be the footprint of the detached parent in the synthetic size, if any?

Comment (stepashka) [reply docs/rfcs/033-ancestor-deletion.md:39]: i expect it to be zero, could you confirm?

Comment (stepashka) [reply docs/rfcs/033-ancestor-deletion.md:39]: ignore this, re-reading this

File: docs/rfcs/033-ancestor-deletion.md [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+  ancestor. If this was the last timeline that depended on the parent, then the
+  parent will become deletable.
+
+This API's implementation will consist of a series of retryable steps, such that
+on failures/timeout it can safely be called again to reach the target state.
+
```
[the remark this answers is not in the document]
Comment (koivunej) [reply docs/rfcs/033-ancestor-deletion.md]: These preconditions have been lifted.

**Comment by koivunej** [comment]
> #### Failures on Postgres 14
> 
> * `test_ancestor_detach_branched_from[True-False-earlier]`: [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-6888/9858213178/index.html#suites/a1c2be32556270764423c495fad75d47/b705e98c155865e7)

Funnily enough, this is another instance of #7830, being fixed already.
